### PR TITLE
Add support for kafka 3 with zookeeper support (adding images for 3.0.1, 3.1.1 and 3.2.0)

### PR DESCRIPTION
**Pull request by lunarfs**
- What is it good for
Support for Kafka 3.0.1, 3.1.1 and 3.2.0 with the regular ZooKeeper setup.

- What I did
@jack12816  did some approach of getting to kafka 3.0.0 I have worked with his commits as basis for getting to 3.0.1,2.1.1 and 3.2.0
the PR is inspired by  https://github.com/wurstmeister/kafka-docker/pull/688  and https://github.com/wurstmeister/kafka-docker/pull/691  I  had to change some Kafka connection options (--zookeeper is deprecated since multiple versions, and since Kafka 3.0 it was removed -- replaced by the --bootstrap-server option. I did a  fix for some tests for the new version/outputs
I have added code to report the no longer supported KAFKA_ADVERTISED_HOST_NAME KAFKA_ADVERTISED_PORT and KAFKA_PORT and KAFKA_HOST_NAME (see   https://github.com/apache/kafka/pull/10872 for more information).
I have updated README.md to reflect that < 3.0 no longer supports  KAFKA_ADVERTISED_HOST_NAME etc.
So i have done some changes to start-kafka.sh to reflect the removed options.. it is a syntactic nightmare of if then else.. and there might be some edge cases i have not yet covered.. feel free to comment 

- A picture of a cute animal (not mandatory but encouraged)
IMG_3708

### CODE REVIEW

**Comment by jimbogithub** [comment]
@lunarfs Like #707 I'm a +1 for adding `3.0.1` support (`3.1.0` doesn't have the bug fix I need).  Could you add that as well?

**Comment by lunarfs** [comment]
> @lunarfs Like #707 I'm a +1 for adding `3.0.1` support (`3.1.0` doesn't have the bug fix I need). Could you add that as well?

I would be happy to supply 3.0.1 support, but I think I would like @wurstmeister to comment a bit on the format of how the approach to the 3.x is done in this PR, is it ok to do it like this or should we do some other approach with the tests. are there new corner cases we need to test.

File: test/0.0/test.start-kafka-bug-313-kafka-opts.kafka.sh [hunk 1]
```diff
@@ -19,5 +19,13 @@ testKafkaOpts() {
 
     echo "  > Set KAFKA_OPTS=$KAFKA_OPTS"
 }
+source "/usr/bin/versions.sh"
 
-testKafkaOpts
+# since 3.0.0 there is no --zookeeper option anymore, so we have to use the
+# --bootstrap-server option with a random broker
+if [[ "$MAJOR_VERSION" -ge "3" ]]; then
+    echo "this thes is obsolete with kafka from version 3.0.0 'advertised.host.name' are removed with Kafka 3.0.0"
```
[the remark this answers is not in the document]
Comment (lunarfs) [reply test/0.0/test.start-kafka-bug-313-kafka-opts.kafka.sh]: thanks, pushed fix

File: create-topics.sh [hunk 2]
```diff
@@ -33,6 +33,14 @@ if [[ "$MAJOR_VERSION" == "0" && "$MINOR_VERSION" -gt "9" ]] || [[ "$MAJOR_VERSI
     KAFKA_0_10_OPTS="--if-not-exists"
 fi
 
+# since 3.0.0 there is no --zookeeper option anymore, so we have to use the
+# --bootstrap-server option with a random broker
+if [[ "$MAJOR_VERSION" -ge "3" ]]; then
+    CONNECT_OPTS="--bootstrap-server $(echo "${BROKER_LIST}" | cut -d ',' -f1)"
```
[the remark this answers is not in the document]
Comment (lunarfs) [reply create-topics.sh:39]: yes; I am just using the first in the list here, do we just want the entire list like 
`CONNECT_OPTS="--bootstrap-server $(echo "${BROKER_LIST}")"
 ?
`

[the remark this answers is not in the document]
Comment (OneCricketeer) [reply create-topics.sh:39]: Don't need to echo anything can simply be `"--bootstrap-server ${BROKER_LIST}"`

[the remark this answers is not in the document]
Comment (lunarfs) [reply create-topics.sh]: ok, updated. 
So in general this PR  turned out to be a bit more complicated than first anticipated..  I will update the description to reflect the changes i did

File: test/scenarios/jmx/docker-compose-kafka3.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+version: '2'
+services:
+  kafka:
+    environment:
+      KAFKA_LISTENERS: PLAINTEXT://:9092
+      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka_1:9092
```
Comment (OneCricketeer) [inline test/scenarios/jmx/docker-compose-kafka3.yml:6]: This should be the hostname (service) of the container

```suggestion
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:9092
```

File: test/scenarios/jmx/docker-compose-kafka.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+version: '2'
+services:
+  kafka:
+    environment:
+      KAFKA_ADVERTISED_HOST_NAME: kafka
+      KAFKA_ADVERTISED_PORT: 9092
```
Comment (OneCricketeer) [inline test/scenarios/jmx/docker-compose-kafka.yml:6]: Not sure why this file is needed. These properties are still deprecated in versions less than Kafka3 and can be replaced by `KAFKA_ADVERTISED_LISTENERS`

Comment (lunarfs) [reply test/scenarios/jmx/docker-compose-kafka.yml:6]: Yes, But I wanted to allow people that run versions prior to 3.0.0 to be able to run with their config unchanged.

File: test/scenarios/jmx/docker-compose-kafka3.yml (hunk 3 above)
Comment (lunarfs) [reply test/scenarios/jmx/docker-compose-kafka3.yml]: nice catch, fixed (did not know about commit suggestion.. I do now)

File: test/scenarios/jmx/docker-compose-kafka.yml (hunk 4 above)
Comment (lunarfs) [reply test/scenarios/jmx/docker-compose-kafka.yml]: @OneCricketeer ok, think i got it... pushed a commit  to fix this

**Comment by jimbogithub** [comment]
@lunarfs I'm testing this out in my local environment.  Looks pretty good except for a couple of issues if I have `KAFKA_CREATE_TOPICS` enabled:

1. `KAFKA_PORT` is used by `create-topics.sh` in order to poll for Kafka to be ready, hence I've had to patch out the startup reject of this setting.
2. `create-topics.sh` also requires `BROKER_LIST` for the 3.x style `CONNECT_OPTS`.  This is easily added to the docker-compose environment, so probably just needs to be documented.

So far so good after making those changes.

**Comment by jimbogithub** [comment]
> @lunarfs I'm testing this out in my local environment. Looks pretty good except for a couple of issues if I have `KAFKA_CREATE_TOPICS` enabled:
> 
> 1. `KAFKA_PORT` is used by `create-topics.sh` in order to poll for Kafka to be ready, hence I've had to patch out the startup reject of this setting.
> 2. `create-topics.sh` also requires `BROKER_LIST` for the 3.x style `CONNECT_OPTS`.  This is easily added to the docker-compose environment, so probably just needs to be documented.
> 
> So far so good after making those changes.

Maybe `BROKER_LIST` could be examined to determine the port instead of re-enabling `KAFKA_PORT`?

**Comment by lunarfs** [comment]
> > @lunarfs I'm testing this out in my local environment. Looks pretty good except for a couple of issues if I have `KAFKA_CREATE_TOPICS` enabled:
> > 
> > 1. `KAFKA_PORT` is used by `create-topics.sh` in order to poll for Kafka to be ready, hence I've had to patch out the startup reject of this setting.
> > 2. `create-topics.sh` also requires `BROKER_LIST` for the 3.x style `CONNECT_OPTS`.  This is easily added to the docker-compose environment, so probably just needs to be documented.
> > 
> > So far so good after making those changes.
> 
> Maybe `BROKER_LIST` could be examined to determine the port instead of re-enabling `KAFKA_PORT`?

Thanks, this was a nice catch, I did not properly modify the tests to catch this scenario.. I have modified tests such that i could capture the scenario. I think the BROKER_LIST is not a valid source for this, but the KAFKA_LISTNERS has the internal port captured, so getting it from there should be fine I think.. and then we also do not need to give extra params in the compose file as KAFKA_LISTNERS is already available  in create-topics.sh

**Comment by lunarfs** [comment]
> > @lunarfs I'm testing this out in my local environment. Looks pretty good except for a couple of issues if I have `KAFKA_CREATE_TOPICS` enabled:
> > 
> > 1. `KAFKA_PORT` is used by `create-topics.sh` in order to poll for Kafka to be ready, hence I've had to patch out the startup reject of this setting.
> > 2. `create-topics.sh` also requires `BROKER_LIST` for the 3.x style `CONNECT_OPTS`.  This is easily added to the docker-compose environment, so probably just needs to be documented.
> > 
> > So far so good after making those changes.
> 
> Maybe `BROKER_LIST` could be examined to determine the port instead of re-enabling `KAFKA_PORT`?

@jimbogithub thanks, this was a nice catch, I did not properly modify the tests to catch this scenario.. I have modified tests such that i could capture the scenario. I think the BROKER_LIST is not a valid source for this, but the KAFKA_LISTNERS has the internal port captured, so getting it from there should be fine I think.. and then we also do not need to give extra params in the compose file as KAFKA_LISTNERS is already available in create-topics.sh.  For the test container however BROKER_LIST is the better option, so if KAFKA_LISTNERS is not available we will use BROKER_LIST.

**Comment by jimbogithub** [comment]
The topic creation is still problematic.  It's not working with the example config:

```
HOSTNAME_COMMAND: curl http://169.254.169.254/latest/meta-data/public-hostname
KAFKA_ADVERTISED_LISTENERS: INSIDE://:9092,OUTSIDE://_{HOSTNAME_COMMAND}:9094
KAFKA_LISTENERS: INSIDE://:9092,OUTSIDE://:9094
KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: INSIDE:PLAINTEXT,OUTSIDE:PLAINTEXT
KAFKA_INTER_BROKER_LISTENER_NAME: INSIDE
```

I use very similar config myself and it results in:
```
kafka0_1     | Exception in thread "main" org.apache.kafka.common.KafkaException: Failed to create new KafkaAdminClient
kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:538)
kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:470)
kafka0_1     | 	at org.apache.kafka.clients.admin.Admin.create(Admin.java:133)
kafka0_1     | 	at kafka.admin.TopicCommand$TopicService$.createAdminClient(TopicCommand.scala:205)
kafka0_1     | 	at kafka.admin.TopicCommand$TopicService$.apply(TopicCommand.scala:209)
kafka0_1     | 	at kafka.admin.TopicCommand$.main(TopicCommand.scala:50)
kafka0_1     | 	at kafka.admin.TopicCommand.main(TopicCommand.scala)
kafka0_1     | Caused by: org.apache.kafka.common.config.ConfigException: Invalid url in bootstrap.servers: OUTSIDE
kafka0_1     | 	at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:59)
kafka0_1     | 	at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:48)
kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:490)
```

Maybe prefer `BROKER_LIST` when available so that it can always be overridden?
Or look for `KAFKA_INTER_BROKER_LISTENER_NAME` to help find the internal address in the `KAFKA_LISTENERS`?

**Comment by dswiecki** [comment]
@lunarfs @jimbogithub are there available any preview docker images of this change?

**Comment by jimbogithub** [comment]
@dswiecki I have an image at jimbodock/kafka:2.13-3.1.1.  It is slightly stripped down (removed the embedded Docker which is only needed for the CI tests).  It is based on the build at the time of https://github.com/wurstmeister/kafka-docker/pull/709#issuecomment-1171028830 so supports `KAFKA_PORT` and `BROKER_LIST` to be able to create topics.

**Comment by mattathompson** [comment]
🙏 Excited for this.

**Comment by lunarfs** [comment]
> The topic creation is still problematic. It's not working with the example config:
> 
> ```
> HOSTNAME_COMMAND: curl http://169.254.169.254/latest/meta-data/public-hostname
> KAFKA_ADVERTISED_LISTENERS: INSIDE://:9092,OUTSIDE://_{HOSTNAME_COMMAND}:9094
> KAFKA_LISTENERS: INSIDE://:9092,OUTSIDE://:9094
> KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: INSIDE:PLAINTEXT,OUTSIDE:PLAINTEXT
> KAFKA_INTER_BROKER_LISTENER_NAME: INSIDE
> ```
> 
> I use very similar config myself and it results in:
> 
> ```
> kafka0_1     | Exception in thread "main" org.apache.kafka.common.KafkaException: Failed to create new KafkaAdminClient
> kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:538)
> kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:470)
> kafka0_1     | 	at org.apache.kafka.clients.admin.Admin.create(Admin.java:133)
> kafka0_1     | 	at kafka.admin.TopicCommand$TopicService$.createAdminClient(TopicCommand.scala:205)
> kafka0_1     | 	at kafka.admin.TopicCommand$TopicService$.apply(TopicCommand.scala:209)
> kafka0_1     | 	at kafka.admin.TopicCommand$.main(TopicCommand.scala:50)
> kafka0_1     | 	at kafka.admin.TopicCommand.main(TopicCommand.scala)
> kafka0_1     | Caused by: org.apache.kafka.common.config.ConfigException: Invalid url in bootstrap.servers: OUTSIDE
> kafka0_1     | 	at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:59)
> kafka0_1     | 	at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:48)
> kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:490)
> ```
> 
> Maybe prefer `BROKER_LIST` when available so that it can always be overridden? Or look for `KAFKA_INTER_BROKER_LISTENER_NAME` to help find the internal address in the `KAFKA_LISTENERS`?

OK, I will take a look at this one of the followin days

**Comment by jozefbarcin** [comment]
:pray:  Excited for this.

**Comment by nise-wg2** [comment]
🙏 Looking forward to this!

**Comment by jimbogithub** [comment]
> > The topic creation is still problematic. It's not working with the example config:
> > ```
> > HOSTNAME_COMMAND: curl http://169.254.169.254/latest/meta-data/public-hostname
> > KAFKA_ADVERTISED_LISTENERS: INSIDE://:9092,OUTSIDE://_{HOSTNAME_COMMAND}:9094
> > KAFKA_LISTENERS: INSIDE://:9092,OUTSIDE://:9094
> > KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: INSIDE:PLAINTEXT,OUTSIDE:PLAINTEXT
> > KAFKA_INTER_BROKER_LISTENER_NAME: INSIDE
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I use very similar config myself and it results in:
> > ```
> > kafka0_1     | Exception in thread "main" org.apache.kafka.common.KafkaException: Failed to create new KafkaAdminClient
> > kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:538)
> > kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:470)
> > kafka0_1     | 	at org.apache.kafka.clients.admin.Admin.create(Admin.java:133)
> > kafka0_1     | 	at kafka.admin.TopicCommand$TopicService$.createAdminClient(TopicCommand.scala:205)
> > kafka0_1     | 	at kafka.admin.TopicCommand$TopicService$.apply(TopicCommand.scala:209)
> > kafka0_1     | 	at kafka.admin.TopicCommand$.main(TopicCommand.scala:50)
> > kafka0_1     | 	at kafka.admin.TopicCommand.main(TopicCommand.scala)
> > kafka0_1     | Caused by: org.apache.kafka.common.config.ConfigException: Invalid url in bootstrap.servers: OUTSIDE
> > kafka0_1     | 	at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:59)
> > kafka0_1     | 	at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:48)
> > kafka0_1     | 	at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:490)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe prefer `BROKER_LIST` when available so that it can always be overridden? Or look for `KAFKA_INTER_BROKER_LISTENER_NAME` to help find the internal address in the `KAFKA_LISTENERS`?
> 
> OK, I will take a look at this one of the followin days

@lunarfs I've added a PR to your PR that resolves the above by simply making the `KAFKA_LISTENERS` parsing smarter.  If you can ripple that through to here then I suspect this is good to go.

**Comment by jimbogithub** [comment]
Preview available here: jimbodock/kafka:2.13-3.3.1.  (with the embedded docker cli stripped out)

**Comment by jozefbarcin** [comment]
> Preview available here: jimbodock/kafka:2.13-3.3.1. (with the embedded docker cli stripped out)

>Works fine for me

**Review by andrewchoi5 — APPROVED** [review]
